File: Java/recover-binary-search-tree.java
/**
 * Problem: https://leetcode.com/problems/recover-binary-search-tree/
 * 
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void recoverTree(TreeNode root) {
        Stack<TreeNode> swapPair = new Stack<TreeNode>();
        recoverTree(root, swapPair);
       
        if (swapPair.size() != 0) {
            TreeNode node1 = swapPair.pop();
            TreeNode node2 = swapPair.pop();
            int val = node1.val;
            node1.val = node2.val;
            node2.val = val;
        }
    }
       
    // In-order Tree Traversal.
    void recoverTree(TreeNode root, Stack<TreeNode> swapPair) {
        if (root == null) return;
        recoverTree(root.left, swapPair);
        addNodeToStack(swapPair, root);
        recoverTree(root.right, swapPair);
    }
   
    // Tricky.
    void addNodeToStack(Stack<TreeNode> swapPair, TreeNode root) {
        int sz = swapPair.size();
        if (sz == 0) {
            swapPair.push(root);
        } else if (sz == 1) {
            if (root.val >= swapPair.peek().val) {
                swapPair.pop();
            }
            swapPair.push(root);
        } else if (sz == 2) {
            if (root.val < swapPair.peek().val) {
                swapPair.pop();
                swapPair.push(root);
            }
        } 
    }
}

File: Swift/recover-binary-search-tree.swift
/**
 * https://leetcode.com/problems/recover-binary-search-tree/
 * 
 * 
 */ 
// Date: Sat Jun 27 11:50:19 PDT 2020
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    
    var firstNode: TreeNode?
    var secondNode: TreeNode?
    var prevNode: TreeNode!
    
    func recoverTree(_ root: TreeNode?) {
        prevNode = TreeNode(Int.min)
        traverse(root)
        let tmpVal = firstNode?.val
        firstNode?.val = secondNode!.val
        secondNode?.val = tmpVal!
    }
    
    private func traverse(_ root: TreeNode?) {
        guard let root = root else { return }
        
        self.traverse(root.left)
        
        if firstNode == nil, prevNode.val >= root.val {
            firstNode = prevNode
        }
        if firstNode != nil, prevNode.val >= root.val {
            secondNode = root
        }
        
        prevNode = root
        
        self.traverse(root.right)
    }
}